Provide PHP 8.1 compatibility

matbech opened the pull request:
This silences the deprecated warnings in PHP 8.1:

- It creates polyfills for each of the `ArrayObject`, `FastPriorityQueue`, `Parameters`, `PriorityList`, `PriorityQueue`, `SplQueue`, and `SplStack` implementations. The new "legacy" implementations keep the same functionality as was present in previous versions, while the "PHP 8.1" implementations add typehints that are required in order to remove deprecation warnings under PHP 8.1 (PHP 8.1 deprecates the ability to define extensions of internal classes that change signatures in any way, even if they fulfill the contravariance/covariance rules.)
  - Autoloading rules alias the appropriate implementation to the original class name based on the PHP version in use.
- It adds `__unserialize()` and `__serialize()` implementations to all classes that previously implemented the `Serializable` interface. The `Serializable` implementations now consume these new methods to make them forwards compatible.
- Because of the polyfill implementations, a large number of code improvements were created to fix new issues flagged by Psalm due to the new class names. In most cases, these are adding type checking to ensure code will work correctly.

### BC concerns

The PHP 8.1 implementations add typehints to signatures where they did not previously exist. In each case, these are compatible with the previous implementation, but they _do_ pose breakage for anybody extending the classes and overriding the relevant methods. **However**, these would lead to deprecation warnings, and, in some cases, fatal errors under PHP 8.1.

Fixes #24

matbech committed:
Implemented magic __serialize and __unserialize methods
This silences the deprecated warnings in PHP 8.1

Signed-off-by: Mathias Berchtold <mberchtold@gmail.com>

matbech committed:
fixes white space

Signed-off-by: Mathias Berchtold <mberchtold@gmail.com>

matbech committed:
replaced tab with spaces

Signed-off-by: Mathias Berchtold <mberchtold@gmail.com>

matbech committed:
fixes more white spaces

Signed-off-by: Mathias Berchtold <mberchtold@gmail.com>

weierophinney committed:
qa: resolve Psalm, CS, and test errors

- Adds more checks during serialization/deserialization for expected types.
- Reuses logic between `Serialization` implementation and magic methods.
- Ensures all CS issues are addressed.
- Fixes a logic error during deserialization of `SplPriorityQueue` (double-decrement error)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: ignore platform req when on PHP 8.1

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
feat: PHP 8.1 support via polyfills

When testing with `error_reporting = E_ALL`, we get deprecation notices that lead to fatal errors when a class implementing an internal interface differs in signature.
In order to resolve those issues, we need to either (a) correct the signature, or (b) add an attribute (e.g. `#[ReturnTypeWillChange]`).
Since both require either changing the code in a BC-impcompatible way, I am proposing we:

- Create a "legacy" and "PHP 8.1" implementation, with the "legacy" implementation being one unchanged from the current implementation, and the "PHP 8.1" implementation performing the requisite signature changes to allow the code to run under that version and above.
- Add an autoloader that performs a `class_alias`, aliasing the appropriate implementation to the original class name.

This approach allows tests to run successfully, and ensures the code runs correctly across versions.
The potential downside is that anybody updating to PHP 8.1 who is extending these classes AND overriding affected methods MAY experience BC breakage.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@matbech I'm making progress on the 8.1 compat, and should have a comprehensive patch to push tomorrow that builds on what you've done here. Thanks for getting this started!

matbech commented:
@weierophinney Thank you for picking this up.

weierophinney committed:
qa: resolve Psalm issues introduced during refactoring

- Fixes unserialize/serialize implementations to ALWAYS call on `__unserialize()` and `__serialize()`.
- Adds type checking wherever possible to remove assumptions.
- Updates Psalm baseline; much of this was already in there under the original classes, and now we've actually resolved old issues.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: updated composer.lock using PHP 7.3

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: exclude PHP 8.1-specific files from CS checks

Due to the fact they raise fatal errors under that version

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
@laminas/technical-steering-committee I'd like a review on this one, as (a) laminas-stdlib is used everywhere, and (b) while I think the approach makes sense (as all potentially non-BC changes are restricted to PHP 8.1, where the code would break without the changes), it _is_ a potential BC break. Additionally, if accepted, it will likely become the template for when we need to introduce PHP 8.1-specific functionality in other components.

Ocramius commented on composer.json:
```diff
@@ -21,8 +21,7 @@
     "require": {
-        "php": "^7.3 || ^8.0",
-        "laminas/laminas-zendframework-bridge": "^1.0"
+        "php": "^7.3 || ^8.0"
```
IMO `^7.3 || ~8.0 || ~8.1` at this point...

boesing replied:
But with the patch `0` as otherwise, tilde works the same way as caret.

samsonasik replied:
should be:

```
^7.3 || ~8.0.0 || ~8.1.0
```

weierophinney replied:
Yes - have this poised to push - just waiting on two more clarifications on other items before I do so.

Ocramius commented on composer.json:
```diff
@@ -53,7 +58,7 @@
     },
-    "replace": {
-        "zendframework/zend-stdlib": "^3.2.1"
+    "conflict": {
```
Thanks for factoring in this change :-)

Ocramius commented on src/ArrayObject/LegacyImplementation.php:
```diff
@@ -33,7 +40,7 @@
  * Extends version-specific "abstract" implementation.
  */
-class ArrayObject implements IteratorAggregate, ArrayAccess, Serializable, Countable
+class LegacyImplementation implements IteratorAggregate, ArrayAccess, Serializable, Countable
```
Could we mark this as `@deprecated` perhaps? People should not reference it directly, IMO

weierophinney replied:
This, and your request to mark the PHP81Implementation classes as `@internal`, is a bit... difficult.

Essentially, these become the actual public API - I'm aliasing the original class name to the appropriate implementation.

If we mark one as "deprecated", then the target class shows up as deprecated when using that version. If we show one as "internal", then the target class shows up as internal for that PHP version.

I think it might need to be a class-level docblock _note_ on each, indicating what each is, and that users should extend the target class instead.

Thoughts?

Ocramius replied:
> Essentially, these become the actual public API - I'm aliasing the original class name to the appropriate implementation.

This is a problem if anyone references the `LegacyImplementation` instead of `ArrayObject`. This **will** happen.

weierophinney replied:
It's a problem for anybody extending the aliased class (in this case, `Laminas\Stdlib\ArrayObject`.

With your proposal:

- Users under PHP versions prior to 8.1 will get deprecation notice due to the `@deprecated` annotation at the class level.
- Users under PHP version 8.1+ will get a notice for using a class marked internal, due to the `@internal` annotation at the class level.

Basically, these _are not_ internal or deprecated - they are the actual public implementations.

Ocramius replied:
Yes, but the symbol exists under this name, and should not be referenced, but it will be referenced.

Ocramius commented on src/ArrayObject/LegacyImplementation.php:
```diff
@@ -236,13 +243,21 @@ public function getFlags()
     /**
      * Create a new iterator from an ArrayObject instance
      *
-     * @return Iterator
```
This change is a bit of a BC break: `Iterator` is more specific than `Traversable`

weierophinney replied:
Question, then: in PHP 8.1, I get an error if the return type is anything other than Traversable, since it's extending/implementing an internal. I can have the legacy implementation stay as it was (`Iterator`), but not the one targeting PHP 8.1.

Should I have `Traversable` as the RTH on the 8.1 implementation, but use an annotation to indicate `Iterator` in that version? (I suspect that will get flagged by Psalm, though...)

Ocramius replied:
Downstream, the expectation is `Iterator`, so we can't just document it differently, then return a generic `Traversable` :-\ 

What is the error if we return `Iterator` here?

weierophinney replied:
There's no error when doing so in the legacy implementation (as we're using docblocks).

It's in the PHP 8.1 version that we get issues. Surprisingly, I don't get them when running phpunit - I see them when running phpcs, where I get the following:

```text
Fatal error: During inheritance of IteratorAggregate: Uncaught PHP_CodeSniffer\Exceptions\RuntimeException: Return type of Laminas\Stdlib\ArrayObject\LegacyImplementation::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in src/ArrayObject/LegacyImplementation.php on line 248 in vendor/squizlabs/php_codesniffer/src/Runner.php:606
```

This won't be an issue until we start running phpcs against PHP 8.1 - but at that time, it will fail, unless there is a change in phpcs.

I'm actually thinking this might be a phpcs issue, with them being too strict in the interpretation - it seems PHP is fine with the change, as it follows the engine's covariance rules. So, should I:

- Update the PHP 8.1 implementation to return `Iterator`, **AND**
- Send an issue upstream to the phpcs project so they can address in their own tooling, **OR**
- leave it as-is

?

Ocramius replied:
The correct solution should be to have `#[ReturnTypeWillChange]`, which we can then remove (and change the return type) in a new major release

Ocramius commented on src/Parameters/LegacyImplementation.php:
```diff
@@ -104,7 +105,7 @@ public function get($name, $default = null)
      * @param string $name
      * @param mixed $value
-     * @return Parameters
+     * @return $this
```
`@return static` - `$this` is not really a type

Ocramius commented on src/Parameters/PHP81Implementation.php:
```diff
@@ -0,0 +1,114 @@
+use function http_build_query;
+use function parse_str;
+
+class PHP81Implementation extends ArrayObject implements ParametersInterface
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/Parameters/LegacyImplementation.php:
```diff
@@ -2,14 +2,15 @@
 use function parse_str;
 
-class Parameters extends PhpArrayObject implements ParametersInterface
+class LegacyImplementation extends ArrayObject implements ParametersInterface
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/FastPriorityQueue/PHP81Implementation.php:
```diff
@@ -0,0 +1,431 @@
+ * the elements. This behaviour can be used in mixed scenarios with high
+ * performance boost.
+ */
+class PHP81Implementation implements Iterator, Countable, Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/FastPriorityQueue/LegacyImplementation.php:
```diff
@@ -27,69 +35,108 @@
  * performance boost.
  */
-class FastPriorityQueue implements Iterator, Countable, Serializable
+class LegacyImplementation implements Iterator, Countable, Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/ArrayObject/PHP81Implementation.php:
```diff
@@ -0,0 +1,495 @@
+ *
+ * Extends version-specific "abstract" implementation.
+ */
+class PHP81Implementation implements IteratorAggregate, ArrayAccess, Serializable, Countable
```
Should be `@internal` IMO: the implementation should not be part of the public API

boesing replied:
when `class_alias` is used, psalm would assume the aliased class as internal as well as it cannot make a difference afaik. We might check that before marking it internal.

Ocramius replied:
Agreed, that is indeed a problem :thinking:

Ocramius replied:
Alternative (ugly, but perhaps acceptable): copy the class to its final location (& rename) during pre-dump-autoload :D

Ocramius commented on src/PriorityList/LegacyImplementation.php:
```diff
@@ -225,9 +230,7 @@ public function key()
-        $node = next($this->items);
-
-        return $node ? $node['data'] : false;
+        next($this->items);
```
The fact that this no longer return smells like BC break

weierophinney replied:
Yes and no.

`Iterator::next()` is documented as being `return void`, and the return value is ignored; it determines if the call to `next()` produces a value by calling `valid()` - and that method has this logic (and always did).

Basically, we were breaking the contract established by the interface previously, and the return value from this was never intended to be used. I consider this a programming mistake instead. (And I likely produced it myself...)

Ocramius replied:
Yes, but then anyone explicitly calling `->next()` (not within a `foreach ()`) would potentially have a breakage here.

weierophinney replied:
I can do one of the following:

- Revert the change. This will require (a) documenting the return value (we didn't before, and instead inherited from the `Iterator` interface... so it was _documented_ as `void`), and (b) marking the return value as deprecated. The question is then what to do about the PHP 8.1 implementation, which _must_ be `void`; we'll have different behavior between the two versions at that point. Additionally, I'll likely need to suppress some additional Psalm errors, due to the difference in the signatures between the interface and implementation.
- Keep the change. The question will be whether or not we call it out, as, again, the previously documented return value was `void`, which would have implied that the return value did not matter.

Knowing that information, what is your vote?

Ocramius commented on src/PriorityList/PHP81Implementation.php:
```diff
@@ -0,0 +1,281 @@
+use function reset;
+use function uasort;
+
+class PHP81Implementation implements Iterator, Countable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/PriorityQueue/LegacyImplementation.php:
```diff
@@ -27,7 +34,7 @@
  * the actual iteration.
  */
-class PriorityQueue implements Countable, IteratorAggregate, Serializable
+class LegacyImplementation implements Countable, IteratorAggregate, Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/PriorityQueue/PHP81Implementation.php:
```diff
@@ -0,0 +1,403 @@
+ * "inner" iterator in the form of an SplPriorityQueue object for performing
+ * the actual iteration.
+ */
+class PHP81Implementation implements Countable, IteratorAggregate, Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/SplQueue/LegacyImplementation.php:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Serializable version of SplQueue
+ */
+class LegacyImplementation extends SplQueue implements Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/SplQueue/PHP81Implementation.php:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Serializable version of SplQueue
+ */
+class PHP81Implementation extends SplQueue
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/SplStack/LegacyImplementation.php:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Serializable version of SplStack
+ */
+class LegacyImplementation extends SplStack implements Serializable
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/SplStack/PHP81Implementation.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Serializable version of SplStack
+ */
+class PHP81Implementation extends SplStack
```
Should be `@internal` IMO: the implementation should not be part of the public API

Ocramius commented on src/autoload.php:
```diff
@@ -0,0 +1,65 @@
+
+// Laminas\Stdlib\ArrayObject
+class_alias(
+    PHP_VERSION_ID >= 80100
```
Can we perhaps rewrite this so that we have one `if` block for `80100`, and then `class_alias()` with `PHP81` vs `Legacy` in each branch (rather than all conditionals here)

kelunik replied:
Why even class aliases instead of conditional definitions / includes?

weierophinney replied:
Conditional definitions are prohibited via our CS checker, and can be brittle. Conditional includes are not a great idea as they (a) have filesystem overhead, and (b) can lead to autoloading conflicts when generating an optimized autoloader (as those will scan all files in a tree to create a map of classes to filenames).

Aliases have the benefit of answering to both the target class and the alias, and fulfill both inheritance trees, allowing slipstreaming them in like this.

Ocramius commented on test/ArrayUtilsTest.php:
```diff
@@ -10,7 +10,7 @@
 use Laminas\Stdlib\ArrayUtils\MergeReplaceKeyInterface;
 use Laminas\Stdlib\Exception\InvalidArgumentException;
-use Laminas\Stdlib\Parameters;
+use laminas\stdlib\parameters;
```
Seems wrong :thinking: 

A lot of these changes in the test suite seem to be overall broken too

weierophinney replied:
Got all of these fixed with 50c8715

weierophinney committed:
fix: pin to PHP 8.x versions using ~ constraint

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: revert change to getIterator() signature in legacy implementation of ArrayObject

Should return `Iterator`, not `Traversable` (PHP 8.1 requires `Traverable`, however).

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: return static, not return $this

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
refactor: use conditionals per version break for autoloading rules

Reduces conditionals from N*2 to 2 (where N was the number of polyfills).

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: fix issues with invalid imports due to usage of class aliases

One of our CS rules assumes that if a class isn't explicitly defined, it does not exist, and therefor references to it should be lowercased.
This causes invalid imports, and leads to odd code errors.
This patch adds rules to exclude that rule when using aliased classes to ensure the import statements are not rewritten.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: revert change to PHP 8.1 implementation of ArrayObject::getIterator

- Set the return value to Iterator
- Provide the `#[ReturnTypeWillChange]` attribute as a hint to static analysis

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Lewiscowles1986 commented on src/ArrayObject/LegacyImplementation.php:
```diff
@@ -240,9 +247,17 @@ public function getFlags()
+        $class    = $this->iteratorClass;
+        $iterator = new $class($this->storage);
+
+        if (! $iterator instanceof Iterator) {
+            throw new UnexpectedValueException(sprintf(
+                'Iterator of type %s is not Traversable',
+                $class
+            ));
+        }
```
where can storage be set to non iterable? It is typed to array, which is AFAIK always iterable. Is this in-case someone assigns iteratorClass to a non-iterable? Surely typing iteratorClass to be a factory which returns iterator from string would solve?

weierophinney replied:
That would be a BC break, though, as currently we allow providing a class _string_ that is then used to instantiate the internal iterator.

Can definitely consider it for a new major version, but for this one, we  just need to ensure type safety.

Lewiscowles1986 commented on src/ArrayObject/LegacyImplementation.php:
```diff
@@ -345,7 +360,17 @@ public function offsetUnset($key)
     public function serialize()
     {
-        return serialize(get_object_vars($this));
+        return serialize($this->__serialize());
+    }
+
+    /**
+     * Magic method used for serializing of an instance.
+     *
+     * @return array
+     */
+    public function __serialize()
+    {
+        return get_object_vars($this);
```
👍

nikic commented:
The intended way to handle this problem is to annotate methods with the `#[ReturnTypeWillChange]` attribute, and then only change the signatures when you are able to (i.e. your next major version).

What is the motivation for not following that approach, and duplicating these class declarations instead? This seems like a lot of complication for unclear gain.

> (PHP 8.1 deprecates the ability to define extensions of internal classes that change signatures in any way, even if they fulfill the contravariance/covariance rules.)

I don't get what this refers to. Extensions of internal classes *should* be following normal variance rules.

weierophinney commented:
> The intended way to handle this problem is to annotate methods with the `#[ReturnTypeWillChange]` attribute, and then only change the signatures when you are able to (i.e. your next major version).
> 
> What is the motivation for not following that approach, and duplicating these class declarations instead? This seems like a lot of complication for unclear gain.

It started because I was first getting CS failures (as in the tool just failed completely due to fatal errors), and then a whole bunch of new Psalm errors (due to signature differences), and then failure to run the test suite across all supported versions (due, again, to errors thrown because of deprecations being raised).

When running our test suite, we have error reporting report deprecations so we can spot issues like this early and adapt; it prevents issues later once deprecations in the engine are actually removed.

Essentially, I needed solutions that would work across all supported versions, while simultaneously ensuring that our code won't needlessly break in the future.

> > (PHP 8.1 deprecates the ability to define extensions of internal classes that change signatures in any way, even if they fulfill the contravariance/covariance rules.)
> 
> I don't get what this refers to. Extensions of internal classes _should_ be following normal variance rules.

This was me misreading errors as I was trying to resolve everything; I later discovered that they _do_ follow normal covariance/contravariance rules. So please ignore that statement.

boesing commented:
I would prefer keeping changes at a minimum rather than duplicating code just to fulfill php upgrade requirements. if the codestyle tool fails, we might switch that rather than changing code in that way.
if psalm has issues, we should find a way to make it understand annotations AND attributes (are we the only framework which uses psalm with docblock and attributes?). 

So I would have expect to add a few attributes and we are ready for 8.1 but this escalated quickly. if all of this is needed to make our tools happy, thats odd

weierophinney commented:
> I would prefer keeping changes at a minimum rather than duplicating code just to fulfill php upgrade requirements. if the codestyle tool fails, we might switch that rather than changing code in that way.

I'll see if I can develop an alternative patch later today.

weierophinney commented:
Closing in favor of #35 , which provides a simpler approach.
